File: python/day4.py
from utilities import read_input

required_fields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid']


def validate_passport(passportInfo, quickly=True):
    passportInfo = passportInfo.replace('\n', ' ')
    passportInfo = passportInfo.split(' ')
    passport = {}

    for info in passportInfo:
        [key, value] = info.split(':')
        passport[key] = value

    for field in required_fields:
        if field not in passport:
            return False

    if quickly:
        return True

    if int(passport['byr']) < 1920 or int(passport['byr']) > 2002:
        return False

    if int(passport['iyr']) < 2010 or int(passport['iyr']) > 2020:
        return False

    if int(passport['eyr']) < 2020 or int(passport['eyr']) > 2030:
        return False

    try:
        height = int(passport['hgt'][:-2])
    except:
        return False
    if 'cm' in passport['hgt']:
        if height < 150 or height > 193:
            return False
    else:
        if height < 59 or height > 76:
            return False

    if len(passport['hcl']) != 7 or passport['hcl'][0] != '#':
        return False
    for c in passport['hcl'][1:]:
        if c.isdigit() or c in 'abcdef':
            continue
        return False

    if passport['ecl'] not in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']:
        return False

    if len(passport['pid']) != 9 or not passport['pid'].isdigit():
        return False

    return True


def part_1(data):
    return sum(map(validate_passport, data))


def part_2(data):
    return sum([validate_passport(passport, False) for passport in data])


if __name__ == '__main__':
    print('Day 4')
    data = read_input('../inputs/day4.txt')
    passports = data.split('\n\n')
    print('\tPart 1: {}'.format(part_1(passports)))
    print('\tPart 2: {}'.format(part_2(passports)))

File: rust/src/day4.rs
use crate::utilities::read_block;
use std::collections::HashMap;

struct Passport {
    byr: i32,
    iyr: i32,
    eyr: i32,
    hgt: String,
    hcl: String,
    ecl: String,
    pid: String,
}

impl Passport {
    fn new(passport_info: String) -> Result<Passport, ()> {
        let passport_info = passport_info.replace('\n', " ");
        let infos: Vec<&str> = passport_info.split_ascii_whitespace().collect();
        let mut map = HashMap::new();
        for info in infos {
            let mut pair = info.split(':');
            let key = pair.next().unwrap();
            let value = pair.next().unwrap();
            map.insert(key.to_string(), value.to_string());
        }
        Ok(Passport {
            byr: map.get("byr").ok_or(())?.parse().unwrap(),
            iyr: map.get("iyr").ok_or(())?.parse().unwrap(),
            eyr: map.get("eyr").ok_or(())?.parse().unwrap(),
            hgt: map.get("hgt").ok_or(())?.to_string(),
            hcl: map.get("hcl").ok_or(())?.to_string(),
            ecl: map.get("ecl").ok_or(())?.to_string(),
            pid: map.get("pid").ok_or(())?.to_string(),
        })
    }

    fn validate(&self) -> bool {
        self.validate_birthdate()
            && self.validate_issue_date()
            && self.validate_expire_date()
            && self.validate_height()
            && self.validate_hair_color()
            && self.validate_eye_color()
            && self.validate_passport_id()
    }

    fn validate_birthdate(&self) -> bool {
        self.byr >= 1920 && self.byr <= 2002
    }

    fn validate_issue_date(&self) -> bool {
        self.iyr >= 2010 && self.iyr <= 2020
    }

    fn validate_expire_date(&self) -> bool {
        self.eyr >= 2020 && self.eyr <= 2030
    }

    fn validate_height(&self) -> bool {
        let height = self.hgt.as_str();
        let len = height.len();
        if len > 3 {
            let unit = &height[len - 2..];
            let height: i32 = height[..len - 2].parse().unwrap();
            if unit == "cm" {
                if height >= 150 && height <= 193 {
                    return true;
                }
            } else if unit == "in" {
                if height >= 59 && height <= 76 {
                    return true;
                }
            }
        }

        false
    }

    fn validate_hair_color(&self) -> bool {
        if self.hcl.len() == 7 {
            let mut hcl = self.hcl.chars();
            if hcl.next().unwrap() == '#' {
                return hcl.all(|c| {
                    c.is_ascii_digit() || (c as usize >= 'a' as usize && c as usize <= 'f' as usize)
                });
            }
        }

        false
    }

    fn validate_eye_color(&self) -> bool {
        // amb blu brn gry grn hzl oth
        let eye_colors = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"];
        for i in 0..eye_colors.len() {
            if self.ecl == eye_colors[i] {
                return true;
            }
        }
        false
    }

    fn validate_passport_id(&self) -> bool {
        self.pid.len() == 9 && self.pid.chars().all(|c| c.is_ascii_digit())
    }
}

fn part_1(passports: &Vec<Passport>) {
    println!("\tPart 1: {}", passports.len());
}

fn part_2(passports: &Vec<Passport>) {
    println!(
        "\tPart 2: {}",
        passports
            .iter()
            .filter(|passport| passport.validate())
            .count()
    );
}

pub fn main() {
    println!("Day 4");
    let passports = read_block("../inputs/day4.txt")
        .into_iter()
        .map(Passport::new)
        .filter(|passport| passport.is_ok())
        .map(|passport| passport.unwrap())
        .collect();
    part_1(&passports);
    part_2(&passports);
}
